File: haskell/pe10.hs
extendPrimeList :: Integer -> [Integer] -> Integer -> Integer -> [Integer]
extendPrimeList n primes intRoot lim
    | lim <= n = primes
    | intRoot * intRoot <= n = extendPrimeList n primes (intRoot + 1) lim
    | null [k | k <- takeWhile (<= intRoot) primes, mod n k == 0] = extendPrimeList (n + 2) (primes ++ [n]) intRoot lim
    | otherwise = extendPrimeList (n+2) primes intRoot lim
    
-- answer = sum $ extendPrimeList 11 [2,3,5,7] 4 2000000
-- this is slow af.

intSqrt n = round $ sqrt n

isPrime :: Integral a => a -> Bool
isPrime n = null [k | k <- [2 .. intSqrt $ fromIntegral n], mod n k == 0]

tillN = 2 * 10^6 :: Integer
answer :: Integer
answer = sum $ filter isPrime [2 .. tillN]
-- this is also slow :(
-- need to figure out a way of fast generation of primes in Haskell

main :: IO ()
main = print answer

File: python/pe10.py
# original algorithm to generate prime list upto a certain integer.
# modified from pe7.py

primes = [2, 3, 5, 7]

def check_primality(n, int_root):
    for p in primes:
        if p > int_root:  # p > n ** 0.5
            break
        elif n % p == 0:  # n is not prime
            return False
    return True

def extend_prime_list(upto_number = 2 * 10 ** 6):
    n = 11
    last_int_root = 4
    last_int = 16  # last_int = last_int_root ** 2
    while n < upto_number:
        if last_int <= n:
            last_int_root += 1
            last_int += 2 * last_int_root - 1
        if check_primality(n, last_int_root):
            primes.append(n)
        n += 2
    return None


if __name__ == "__main__":
    extend_prime_list()
    answer = sum(primes)
    print(answer)
